src/main/java/appeng/parts/p2p/PartP2PIC2Power.java

```diff
@@ -19,247 +19,241 @@
+
+import appeng.api.config.PowerUnits;
+import appeng.integration.IntegrationType;
+import appeng.items.parts.PartModels;
+import appeng.me.GridAccessException;
+import appeng.me.cache.helpers.TunnelCollection;
+import appeng.transformer.annotations.Integration.Interface;
+import appeng.transformer.annotations.Integration.InterfaceList;
+import appeng.util.Platform;
+
+
+@InterfaceList( value = {
+		@Interface( iface = "ic2.api.energy.tile.IEnergySink", iname = IntegrationType.IC2 ),
+		@Interface( iface = "ic2.api.energy.tile.IEnergySource", iname = IntegrationType.IC2 )
+} )
+public class PartP2PIC2Power extends PartP2PTunnel<PartP2PIC2Power> implements IEnergySink, IEnergySource
+{
+
+	private static final P2PModels MODELS = new P2PModels( "part/p2p/p2p_tunnel_ic2" );
+
+	@PartModels
+	public static List<ResourceLocation> getModels()
+	{
+		return MODELS.getModels();
+	}
+
+	// two packet buffering...
+	private double OutputEnergyA;
+	private double OutputEnergyB;
+	// two packet buffering...
+	private double OutputVoltageA;
+	private double OutputVoltageB;
+
+	public PartP2PIC2Power( ItemStack is )
+	{
+		super( is );
+	}
+
+	@Override
+	public void readFromNBT( NBTTagCompound tag )
+	{
+		super.readFromNBT( tag );
+		this.OutputEnergyA = tag.getDouble( "OutputPacket" );
+		this.OutputEnergyB = tag.getDouble( "OutputPacket2" );
+		this.OutputVoltageA = tag.getDouble( "OutputVoltageA" );
+		this.OutputVoltageB = tag.getDouble( "OutputVoltageB" );
+	}
+
+	@Override
+	public void writeToNBT( NBTTagCompound tag )
+	{
+		super.writeToNBT( tag );
+		tag.setDouble( "OutputPacket", this.OutputEnergyA );
+		tag.setDouble( "OutputPacket2", this.OutputEnergyB );
+		tag.setDouble( "OutputVoltageA", this.OutputVoltageA );
+		tag.setDouble( "OutputVoltageB", this.OutputVoltageB );
+	}
+
+	@Override
+	public void onTunnelConfigChange()
+	{
+		this.getHost().partChanged();
+	}
+
+	@Override
+	public void onTunnelNetworkChange()
+	{
+		this.getHost().notifyNeighbors();
+	}
+
+	@Override
+	public boolean acceptsEnergyFrom( IEnergyEmitter emitter, EnumFacing direction )
+	{
+		return !this.isOutput() && direction == this.getSide().getFacing();
+	}
+
+	@Override
+	public boolean emitsEnergyTo( IEnergyAcceptor receiver, EnumFacing direction )
+	{
+		return this.isOutput() && direction == this.getSide().getFacing();
+	}
+
+	@Override
+	public double getDemandedEnergy()
+	{
+		if( this.isOutput() )
+		{
+			return 0;
+		}
+
+		try
+		{
+			for( PartP2PIC2Power t : this.getOutputs() )
+			{
+				if( t.OutputEnergyA <= 0.0001 || t.OutputEnergyB <= 0.0001 )
+				{
+					return 2048;
+				}
+			}
+		}
+		catch( GridAccessException e )
+		{
+			return 0;
+		}
+
+		return 0;
+	}
+
+	@Override
+	public int getSinkTier()
+	{
+		return 4;
+	}
+
+	@Override
+	public double injectEnergy( EnumFacing directionFrom, double amount, double voltage )
+	{
+		TunnelCollection<PartP2PIC2Power> outs;
+		try
+		{
+			outs = this.getOutputs();
+		}
+		catch( GridAccessException e )
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)